cmds/mem.c

```diff
@@ -0,0 +1,284 @@
+/*
+ * Phoenix-RTOS
+ *
+ * MEMory read/write tool
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <hal/hal.h>
+#include <lib/lib.h>
+#include <syspage.h>
+#include "cmd.h"
+
+
+static void cmd_memInfo(void)
+{
+	lib_printf("Reads or writes values from/to a range of memory addresses");
+}
+
+
+static void cmd_memUsage(void)
+{
+	lib_printf(
+		"Usage: mem [options].. address\n"
+		"\t-F        Disable validation of memory (map) regions\n"
+		"\t-a align  Memory alignment (default=%u)\n"
+		"\t-s size   Size of write/read (1,2,4,8)\n"
+		"\t-c count  Number of repetitions with address step of (size)\n"
+		"\t-w value  Write (value) to memory address\n"
+		"\t-r        Read memory from address\n",
+		(unsigned int)sizeof(addr_t));
+}
+
+
+static int region_validator(addr_t addr, size_t size, unsigned int attr)
+{
+	unsigned int mapAttr = 0uL;
+
+	if (syspage_mapRangeCheck(addr, addr + size, &mapAttr) == 0) {
+		return 0;
+	}
+
+	return ((mapAttr & attr) == attr) ? 1 : 0;
+}
+
+
+static int memr(addr_t addr, size_t size, int forceValid)
+{
+	union {
+		u64 val64;
+		u32 val32;
+		u16 val16;
+		u8 val8;
+	} u = { .val64 = 0uLL };
+
+	if (sizeof(addr_t) <= 8uLL) {
+		lib_printf("read 0x%08x = ", addr);
+	}
+	else {
+		lib_printf("read 0x%016llx = ", (u64)addr);
+	}
```

Comment:
```suggestion
	lib_printf("read 0x%p = ", (void *)addr);
```

Reply:
I've been here already :), `0x%p` for `0` address returns weird things.

![2023-08-07-093722_1889x86_scrot](https://github.com/phoenix-rtos/plo/assets/141153/86557bf9-1405-45ae-8e31-d55ccd384da9)

Reply:
Then maybe NULL check only? Accessing NULL is rare use-case anyway

Reply:
Ok, changed to `%p` version.